[0.6.4] Use link local address only if necessary fix #292

G10h4ck opened the pull request:
This should solve APIPA address being always selected as preferred local
address on Windows which causes problems connecting to machines on
the same local network.
Now the link local address is used as fallback only if there is non link
local private address available.

G10h4ck committed:
Use link local address only if necessary

This should solve APIPA address being always selected as preferred local
address on Windows which was causing problems connecting to machines on
the same local network.
Now the link local address is used as fallback only if there is non link
local private address available.

G10h4ck commented:
this should fix https://github.com/RetroShare/RetroShare/issues/292

csoler commented:
Could you please replace the "for(auto&& addrs: something)" by a for(auto it(something.begin();it!=something.end();++it), for backward compatibility.

csoler commented:
what do you mean by "Having just one local address is a flawed assumption" ? getLocalAddresses() mainly retrieves a network interface, which comes with its local address. And RS only talks to one network interface.

G10h4ck committed:
Use retrocompatible iterators

Explain why shuffling the local addresses list before looking for
 a viable local is a bad idea.

G10h4ck committed:
Improve discovery

Workaround RS single local address flawed assumption by spreading
  all viables local addrresses via discovery
Spread addresses also through hidden node as they are the only hope
  to help direct connection happening for common trusted nodes
  behind nasty firewalls

G10h4ck committed:
Increase size of list of ip shared via discovery2

G10h4ck committed:
Prevent discovery to share local addresses if hidden

Do not query/share local addresses if hidden node as it would leak
  information about the location of the node

G10h4ck committed:
Discovery do not distribute addresses to hidden nodes

As there is no consensus in takde advantage of hidden nodes for
  discovery do not spread ip address to them
